stan/math/rev/mat/fun/multiply.hpp

```diff
@@ -1,23 +1,407 @@
 #ifndef STAN_MATH_REV_MAT_FUN_MULTIPLY_HPP
 #define STAN_MATH_REV_MAT_FUN_MULTIPLY_HPP
 
-#include <stan/math/rev/core.hpp>
-#include <stan/math/rev/mat/fun/Eigen_NumTraits.hpp>
-#include <stan/math/rev/mat/fun/typedefs.hpp>
+#include <boost/utility/enable_if.hpp>
+#include <boost/type_traits.hpp>
 #include <stan/math/prim/mat/fun/Eigen.hpp>
 #include <stan/math/prim/mat/fun/typedefs.hpp>
-#include <stan/math/prim/mat/err/check_multiplicable.hpp>
+#include <stan/math/rev/scal/fun/value_of_rec.hpp>
+#include <stan/math/rev/scal/fun/value_of.hpp>
 #include <stan/math/rev/mat/fun/to_var.hpp>
-#include <stan/math/rev/mat/fun/dot_product.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/prim/mat/fun/value_of_rec.hpp>
+#include <stan/math/prim/mat/err/check_multiplicable.hpp>
+#include <stan/math/prim/scal/err/check_not_nan.hpp>
 #include <boost/utility/enable_if.hpp>
 #include <boost/type_traits.hpp>
 #include <boost/math/tools/promotion.hpp>
 
 namespace stan {
   namespace math {
 
+    template <typename TA, int RA_, int CA_, typename TB, int CB_>
+    class multiply_mat_vari : public vari {
+    public:
+      int RA, CA, CB, Asize, Bsize;  // A.cols() = B.rows()
+      double* Ad_a;
+      double* Bd_a;
+      vari** variRefA_;
+      vari** variRefB_;
+      vari** variRefAB_;
+
+      /* ctor for multiply function
+       *
+       * Stores pointers to varis for A
+       * Stores pointers to varis for B
+       * Stores pointers to varis for AB
+       * All memory allocated on ChainableStack
+       *
+       * @tparam TA Scalar type for matrix A
+       * @tparam RA_ Rows for matrix A
+       * @tparam CA_ Columns for matrix A, Rows for matrix B
+       * @tparam TB Scalar type for matrix B
+       * @tparam CB_ Columns for matrix B
+       * @param matrix A
+       * @param matrix B
+       * */
+     multiply_mat_vari(const Eigen::Matrix<TA, RA_, CA_>& A,
+                       const Eigen::Matrix<TB, CA_, CB_>& B)
+        : vari(0.0),
+          RA(A.rows()), CA(A.cols()),
+          CB(B.cols()), Asize(A.size()), Bsize(B.size()),
+          Ad_a(ChainableStack::memalloc_.alloc_array<double>
+               (A.rows() * A.cols())),
+          Bd_a(ChainableStack::memalloc_.alloc_array<double>
+               (B.rows() * B.cols())),
+          variRefA_(ChainableStack::memalloc_.alloc_array<vari*>
+                    (A.rows() * A.cols())),
+          variRefB_(ChainableStack::memalloc_.alloc_array<vari*>
+                    (B.rows() * B.cols())),
+          variRefAB_(ChainableStack::memalloc_.alloc_array<vari*>
+                    (A.rows() * B.cols())) {
+            for (size_type i = 0; i < A.size(); ++i) {
+                variRefA_[i] = A.coeffRef(i).vi_;
+                Ad_a[i] = A.coeffRef(i).val();
+            }
+            for (size_type i = 0; i < B.size(); ++i) {
+                variRefB_[i] = B.coeffRef(i).vi_;
+                Bd_a[i] = B.coeffRef(i).val();
+            }
```

Comment:
Although counterintuitive, these may be faster broken out into two loops in order to preserve as much memory locality as possible (though here the use of A on the right side complicates matters, or it'd be a no-brainer to use two loops).  I also don't know if `A.coeffRef()` gets inlined or whether it'd be worth assigning to a local to prevent recomputation.  The integer loop calcs will get parallelized in the CPU and won't wind up costing anything.